### PR TITLE
Allow monitoring directories as builder sources

### PR DESCRIPTION
**Pull request by pwithnall**
See the commit messages.

tl;dr: This reduces the number of file monitors in gnome-software by an order of magnitude.

### CODE REVIEW

**Comment by pwithnall** [comment]
This is used in https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/469.

File: src/xb-builder.c [hunk 1]
```diff
@@ -648,11 +648,15 @@ xb_builder_watch_source (XbBuilder *self,
...
 	GFile *file = xb_builder_source_get_file (source);
+	g_autoptr(GFile) to_watch = NULL;
 	if (file == NULL)
 		return TRUE;
-	if ((xb_builder_source_get_flags (source) & XB_BUILDER_SOURCE_FLAG_WATCH_FILE) == 0)
+	if ((xb_builder_source_get_flags (source) & (XB_BUILDER_SOURCE_FLAG_WATCH_FILE | XB_BUILDER_SOURCE_FLAG_WATCH_DIRECTORY)) == 0)
 		return TRUE;
-	if (!xb_silo_watch_file (priv->silo, file, cancellable, error))
+
+	to_watch = (xb_builder_source_get_flags (source) & XB_BUILDER_SOURCE_FLAG_WATCH_DIRECTORY) ? g_file_get_parent (file) : g_object_ref (file);
```
Comment (hughsie) [inline src/xb-builder.c:657]: super long, can this just be an `if` block please

File: src/xb-builder.c [hunk 2]
```diff
@@ -648,11 +648,15 @@ xb_builder_watch_source (XbBuilder *self,
 {
 	XbBuilderPrivate *priv = GET_PRIVATE (self);
 	GFile *file = xb_builder_source_get_file (source);
+	g_autoptr(GFile) to_watch = NULL;
```
Comment (hughsie) [inline src/xb-builder.c:651]: maybe `file_monitor`?

Comment (pwithnall) [reply src/xb-builder.c]: `file_monitor` makes it sound like it’s a `GFileMonitor`. I’ve changed it to `watched_file`, how’s that?

File: src/xb-silo.c [hunk 3]
```diff
@@ -1311,8 +1310,8 @@ static void
 xb_silo_init (XbSilo *self)
 {
 	XbSiloPrivate *priv = GET_PRIVATE (self);
-	priv->file_monitors = g_hash_table_new_full (g_str_hash, g_str_equal,
-						     g_free, (GDestroyNotify) xb_silo_file_monitor_item_free);
+	priv->file_monitors = g_hash_table_new_full (g_file_hash, (GEqualFunc) g_file_equal,
+						     g_object_unref, (GDestroyNotify) xb_silo_file_monitor_item_free);
```
Comment (hughsie) [inline src/xb-silo.c:1314]: Doesn't this need another (GDestroyNotify) to avoid a compile warning?

Comment (pwithnall) [reply src/xb-silo.c:1314]: Nope, `g_object_unref` has the same signature as `GDestroyNotify`. No compile warnings for me here

Comment (hughsie) [reply src/xb-silo.c:1314]: My mistake, apologies.

**Pull request merged**